src/core/qgsjsonutils.cpp

```diff
@@ -119,7 +121,7 @@ QString QgsJSONExporter::exportFeature( const QgsFeature &feature, const QVarian
 
     if ( mIncludeAttributes )
     {
-      QgsFields fields = feature.fields();
+      QgsFields fields = mLayer.data() ? mLayer->fields() : feature.fields();
```

Comment:
Probably `mLayer ? ...` is sufficient (and marginally more performant), guessing that QPointer implements a bool operator.